### ISSUE TITLE
Unable to Reproduce Results!

### ISSUE THREAD

**Issue by SnowXYZ2021** [ASCIIJK/MiN-NeurIPS2025#3]
Hi, @ASCIIJK.

Thank you for your work. I attempted to reproduce your results but encountered significant issues, with the model performing poorly during the inference phase. I have verified that I installed the recommended environments, loaded the correct pre-trained models, and made no modifications to the source code. Additionally, I confirmed that the dataset is not the issue, as I was able to reproduce other works (e.g., EASE) using my data. I noticed that you used custom evaluation code instead of the code provided in PILOT, which led me to suspect that the evaluation code might be the problem.

Could you please check this issue? May I ask if you have attempted to validate the code by cloning the repository via Git?

Looking forward to your prompt response.

```
2025-11-07 16:50:02,961 [BaseTrainer.py] => dataset: imagenetr
2025-11-07 16:50:02,962 [BaseTrainer.py] => init_class: 20
2025-11-07 16:50:02,962 [BaseTrainer.py] => increment: 20
2025-11-07 16:50:02,962 [BaseTrainer.py] => img_size: 224
2025-11-07 16:50:02,962 [BaseTrainer.py] => save_all_checkpoint: False
2025-11-07 16:50:02,962 [BaseTrainer.py] => shuffle: True
2025-11-07 16:50:02,962 [BaseTrainer.py] => seed: 1993
2025-11-07 16:50:02,962 [BaseTrainer.py] => model: min
2025-11-07 16:50:02,962 [BaseTrainer.py] => input_size: 224
2025-11-07 16:50:02,962 [BaseTrainer.py] => device: cuda:0
2025-11-07 16:50:02,962 [BaseTrainer.py] => backbone_type: pretrained_vit_b16_224_in21k_min
2025-11-07 16:50:02,962 [BaseTrainer.py] => pretrained: True
2025-11-07 16:50:02,962 [BaseTrainer.py] => memery_size: 0
2025-11-07 16:50:02,962 [BaseTrainer.py] => memery_per_class: 0
2025-11-07 16:50:02,962 [BaseTrainer.py] => optimizer_type: sgd
2025-11-07 16:50:02,962 [BaseTrainer.py] => scheduler_type: step
2025-11-07 16:50:02,962 [BaseTrainer.py] => init_epochs: 10
2025-11-07 16:50:02,962 [BaseTrainer.py] => init_lr: 0.001
2025-11-07 16:50:02,962 [BaseTrainer.py] => init_batch_size: 128
2025-11-07 16:50:02,962 [BaseTrainer.py] => init_weight_decay: 0.0005
2025-11-07 16:50:02,962 [BaseTrainer.py] => epochs: 10
2025-11-07 16:50:02,962 [BaseTrainer.py] => lr: 0.001
2025-11-07 16:50:02,962 [BaseTrainer.py] => batch_size: 128
2025-11-07 16:50:02,962 [BaseTrainer.py] => buffer_batch: 1000
2025-11-07 16:50:02,962 [BaseTrainer.py] => fit_epochs: 3
2025-11-07 16:50:02,962 [BaseTrainer.py] => weight_decay: 0.0005
2025-11-07 16:50:02,962 [BaseTrainer.py] => num_workers: 4
2025-11-07 16:50:02,962 [BaseTrainer.py] => hidden_dim: 192
2025-11-07 16:50:02,962 [BaseTrainer.py] => buffer_size: 16384
2025-11-07 16:50:02,962 [BaseTrainer.py] => gamma: 500
2025-11-07 16:50:04,375 [_builder.py] => Loading pretrained weights from Hugging Face hub (timm/vit_base_patch16_224.augreg_in21k)
2025-11-07 16:50:05,714 [MiN.py] => task_list: ['n02099601', 'n02088466', 'n01833805', 'n02091032', 'n03710193', 'n07718472', 'n02410509', 'n04465501', 'n01694178', 'n02481823', 'n02099712', 'n09472597', 'n01518878', 'n03494278', 'n02110341', 'n07880968', 'n03930630', 'n02219486', 'n02607072', 'n02206856']
2025-11-07 16:50:05,714 [MiN.py] => task_order: [168, 136, 51, 9, 183, 101, 171, 99, 42, 159, 191, 70, 16, 188, 27, 10, 175, 26, 68, 187]
2025-11-07 16:50:18,613 [MiN.py] => Task 0 --> Learning Beneficial Noise!: Epoch 1/10 => Loss 2.797, train_accy 15.99
2025-11-07 16:50:24,580 [MiN.py] => Task 0 --> Learning Beneficial Noise!: Epoch 2/10 => Loss 1.971, train_accy 52.94
2025-11-07 16:50:30,642 [MiN.py] => Task 0 --> Learning Beneficial Noise!: Epoch 3/10 => Loss 1.210, train_accy 70.60
2025-11-07 16:50:36,356 [MiN.py] => Task 0 --> Learning Beneficial Noise!: Epoch 4/10 => Loss 0.851, train_accy 79.63
2025-11-07 16:50:42,054 [MiN.py] => Task 0 --> Learning Beneficial Noise!: Epoch 5/10 => Loss 0.687, train_accy 83.65
2025-11-07 16:50:47,915 [MiN.py] => Task 0 --> Learning Beneficial Noise!: Epoch 6/10 => Loss 0.599, train_accy 85.46
2025-11-07 16:50:54,154 [MiN.py] => Task 0 --> Learning Beneficial Noise!: Epoch 7/10 => Loss 0.559, train_accy 85.64
2025-11-07 16:51:00,328 [MiN.py] => Task 0 --> Learning Beneficial Noise!: Epoch 8/10 => Loss 0.505, train_accy 86.90
2025-11-07 16:51:06,494 [MiN.py] => Task 0 --> Learning Beneficial Noise!: Epoch 9/10 => Loss 0.424, train_accy 88.75
2025-11-07 16:51:12,720 [MiN.py] => Task 0 --> Learning Beneficial Noise!: Epoch 10/10 => Loss 0.393, train_accy 89.30
2025-11-07 16:51:21,172 [MiN.py] => Task 0 --> Update Analytical Classifier!
2025-11-07 16:51:26,799 [MiN.py] => Task 0 --> Update Analytical Classifier!
2025-11-07 16:51:32,421 [MiN.py] => Task 0 --> Update Analytical Classifier!
2025-11-07 16:51:39,171 [MiN.py] => Task 0 --> Reupdate Analytical Classifier!
2025-11-07 16:51:40,388 [MiN.py] => Task 0 --> Reupdate Analytical Classifier!
2025-11-07 16:51:40,841 [MiN.py] => Task 0 --> Reupdate Analytical Classifier!
2025-11-07 16:51:43,340 [MiN.py] => total acc: [7.48]
2025-11-07 16:51:43,340 [MiN.py] => avg_acc: 7.48
2025-11-07 16:51:43,341 [MiN.py] => task_confusion_metrix:
[[655.]]
2025-11-07 16:51:43,342 [MiN.py] => task_list: ['n07742313', 'n01531178', 'n02117135', 'n02106662', 'n03481172', 'n07920052', 'n02134084', 'n07768694', 'n02823750', 'n07753592', 'n04310018', 'n02356798', 'n02346627', 'n02363005', 'n03124170', 'n03602883', 'n02097298', 'n04325704', 'n01843383', 'n02085620']
2025-11-07 16:51:43,342 [MiN.py] => task_order: [98, 6, 85, 35, 112, 43, 100, 0, 103, 181, 88, 59, 4, 2, 116, 174, 94, 80, 106, 1]
```

**Comment by ASCIIJK** [comment]
I compared the log files I uploaded with the ones you uploaded, and noticed that the confusion matrix for our first task was different (655 in yours and 689 in mine). Therefore, I guess the problem lies in the test set. I think that maybe the test category and the training category you used were not matched properly, so this result is approximately a random value. (The category index of our training index are the same, but the number of test sets is not the same.) You can try to check the test_dir in 'data_process/data.py'. Maybe can fix this problem.

**Comment by SnowXYZ2021** [comment]
Thank you for your prompt reply! I can confirm that the data is error-free. I think this might be a general issue, because I also encountered a similar problem on the Omnibenchmark dataset.

**Comment by SnowXYZ2021** [comment]
@ASCIIJK May I ask if you used the same data split as in previous works, such as APER? I am puzzled because the first task includes 688 samples, whereas if you used the same dataset provided by APER, the actual number should be **655 instead of 689**. Furthermore, since I can reproduce the results of models, including APER and EASE, using my own dataset, I suspect that there may be some issues with your data and code. 

Given the potential impact on the paper's conclusions, and this paper has been accepted by NeurIPS, I believe this warrants a thorough and rigorous investigation.

**Comment by ASCIIJK** [comment]
I have also noticed this issue. I noticed that the category index was the same but the category name is different in our log files. 

However, I believe the main issue lies in the fact that the categories of the training set and the test set are not matched.

Before uploading this code, I had tested most of the benchmarks, but this issue didn't occur. I'm currently also very confused. This might be related to the sequence in which the system loads the data.

I can ensure data division in my comparative experiment was the same. Because I have compared both their indexes and category names, that's why I printed out the category index. I reproduced the comparison methods on the same device, but did not directly use the data from the original paper. Additionally, I also conducted experiments with multiple sets of random seeds, and the results were basically consistent and stable.

**Comment by SnowXYZ2021** [comment]
Thank you for your continued attention to this matter. I have attempted to use different data splits with both datasets (omnibenchmark and imagenet-r), and the results during inference are consistently below 10%. It appears that the code can only run normally on your own dataset and fails to adapt to any other dataset, which is a critical concern for the generalizability and reproducibility of the work, especially for a high-profile venue like NeurIPS.

To help resolve this, I have two suggestions:

1. Download the dataset used in APER and train your model to identify where the issue lies.
2. Release the datasets used and provide implementation details and logs for the baseline methods, so that we can better understand the discrepancies and ensure the reproducibility of the results.

The authors should be aware that their paper has been accepted to a top conference NeurIPS, and therefore, I hope they will conduct a thorough investigation into this matter and revise the code to provide a fair and accurate comparison. Rather than expressing a curious attitude towards these issues, it is essential to address them promptly and properly. Failing to do so may mislead many readers and potentially damage the authors' academic reputation.

**Comment by ASCIIJK** [comment]
Thanks for your advice. But I have provided the link of IN-R and I downloaded the datasets in APER (PILOT repo). I checked both datasets last minutes, and they are the same. And as I know, someone has successfully run this code and reproduced the results. I am truly confused about your issues.

In addition, your problem seems not be caused by the data splits. I will re-run this code to check for any potential issues.  Thank you very much for your question. I hope you can provide some more detailed information, such as your device details or the structure of your data set. This will help me fix the problem.

**Comment by SnowXYZ2021** [comment]
Thank you for your prompt response. 

I downloaded the dataset from [this link](https://drive.google.com/file/d/1SG4TbiL8_DooekztyCVK8mPmfhMo8fkR/view?usp=sharing), extracted the files, and found the train and test directories. I then modified the [paths in the code](https://github.com/ASCIIJK/MiN-NeurIPS2025/blob/292d8834b92a4b908939241b535d46e4e94d6bfe/MiN/data_process/data.py#L224) to point to these directories and ran the script without making any further changes.

You mentioned that the imagenet-r dataset you used is the same as APER, but I would like to highlight that the test set for the first task should contain 655 samples, not 689 as mentioned in your log. If you counted the samples in the directories, you would find that the actual dataset you used is still different from APER.

I understand that you have provided the link to ImageNet-R, but it does not include the train-test split. My suggestion (point 2 above) is that you should provide the pre-processed data, similar to APER, to better ensure reproducibility.

**Comment by SnowXYZ2021** [comment]
Now, my colleagues have also encountered the same issue, and we have followed the same steps:

1. Downloaded the ImageNet-R dataset from [this link](https://drive.google.com/file/d/1SG4TbiL8_DooekztyCVK8mPmfhMo8fkR/view?usp=sharing) and unzipped it to obtain the train and test folders.
2. Clone the code and modify the [corresponding paths](https://github.com/ASCIIJK/MiN-NeurIPS2025/blob/292d8834b92a4b908939241b535d46e4e94d6bfe/MiN/data_process/data.py#L224) to point to the train and test directories. Apart from this, we haven't made any modifications to the code
3. During the first inference process, we observed an accuracy of less than 10%.


We have reproduced this issue on different devices, including 4090 and A100.

**Comment by ASCIIJK** [comment]
Thank you for your attempt. I am currently reconfiguring the environment to try running it again. I will fix this issue as soon as possible.

**Comment by SnowXYZ2021** [comment]
Thank you for your prompt action. Looking forward to your updates.

**Comment by ASCIIJK** [comment]
<img width="1480" height="595" alt="Image" src="https://github.com/user-attachments/assets/accc68de-f339-4adc-9d63-84a5947b008a" />

I am confused about this problem. I just re-downloaded this code and then re-configured the environment on a new server, using the IN-R data downloaded from the APER repository. This is the screenshot I recreated. I'm not sure where the problem lies anymore...

Perhaps it could be considered to manually import the pre-trained weights like this? Since I am unable to access Hugging Face normally, I need to use a VPN on another computer to download the weights, and then upload them to the server for the experiments. (Although I think it's highly unlikely to be this problem...)
<img width="1426" height="91" alt="Image" src="https://github.com/user-attachments/assets/5c1eca06-5c92-44d2-9ff5-a3b63603adc9" />

I'm not quite sure where the problem lies either. Maybe you could show me the training and testing paths in the 'data.py'. Maybe I can identify some issues then. Currently, I can reproduce this code on an ordinary server (This is the server I just rented. Its GPU is 4090.).

**Comment by SnowXYZ2021** [comment]
Thank you for your kind reply.

The core issue lies in the ``split_img_label(root)`` function, which uses ``os.listdir(root)`` to retrieve the list of class directories.

The crucial problem is that ``os.listdir()`` returns items in an arbitrary, filesystem-dependent order. This order is not guaranteed to be alphabetical and can differ between consecutive calls on different directories. This inconsistency can lead to incorrect class labels and directory mappings, ultimately affecting the model's performance. 

My suggestion is to strictly use the same style code of PIOLT to ensure the class order is the same as the previous works.

**Comment by SnowXYZ2021** [comment]
By using the following code, I observed different orders between the training and test folders. However, after adding the sorted function to ensure consistent ordering, the returned lists are identical.
```
def split_img_label(root):
    category_index = os.listdir(root)
    cat_list = []
    num = 0
    print("start loading the img index!")
    for cat in category_index:
        sample_list = os.listdir(os.path.join(root, cat))
        sample_list = [(os.path.join(root, cat, sample_list[i]), num) for i in range(len(sample_list))]
        cat_list.append(sample_list)
        num += 1
    return category_index, cat_list

root = '/data/dataset/imagenet-r/train/'
category_index_train, train_data = split_img_label(root)

root = '/data/dataset/imagenet-r/test/'
category_index_test, test_data = split_img_label(root)

print("Checking class order consistency between train and test sets...")
assert category_index_train == category_index_test, "Class orders do not match!"
```

**Comment by ASCIIJK** [comment]
Thank you for your suggestion. I did overlook the issue of folder sorting across different systems. Thank you again, and I will fix this issue as soon as possible.
